source/common/network/listen_socket_impl.h

```diff
@@ -56,23 +56,52 @@ class ListenSocketImpl : public SocketImpl {
   ListenSocketImpl(int fd, const Address::InstanceConstSharedPtr& local_address)
       : SocketImpl(fd, local_address) {}
 
+  void setupSocket(const Network::Socket::OptionsSharedPtr& options, bool bind_to_port);
   void doBind();
   void setListenSocketOptions(const Network::Socket::OptionsSharedPtr& options);
 };
 
 /**
  * Wraps a unix socket.
  */
-class TcpListenSocket : public ListenSocketImpl {
+template <Address::SocketType T> struct NetworkSocketTrait {};
+
+template <> struct NetworkSocketTrait<Address::SocketType::Stream> {
+  static constexpr Address::SocketType type = Address::SocketType::Stream;
+};
+
+template <> struct NetworkSocketTrait<Address::SocketType::Datagram> {
+  static constexpr Address::SocketType type = Address::SocketType::Datagram;
+};
+
+template <typename T> class NetworkListenSocket : public ListenSocketImpl {
 public:
-  TcpListenSocket(const Address::InstanceConstSharedPtr& address,
-                  const Network::Socket::OptionsSharedPtr& options, bool bind_to_port);
-  TcpListenSocket(int fd, const Address::InstanceConstSharedPtr& address,
-                  const Network::Socket::OptionsSharedPtr& options);
+  NetworkListenSocket(const Address::InstanceConstSharedPtr& address,
+                      const Network::Socket::OptionsSharedPtr& options, bool bind_to_port)
+      : ListenSocketImpl(address->socket(T::type), address) {
+    RELEASE_ASSERT(fd_ != -1, "");
+
+    setProtocolSpecificSocketOptions();
+
+    setupSocket(options, bind_to_port);
+  }
+
+  NetworkListenSocket(int fd, const Address::InstanceConstSharedPtr& address,
```

Comment:
Don't we need to call `setProtocolSpecificSocketOptions()` in this variant also? Can we somehow unify the setup logic? It's a little confusing as with the constructor variants, protocol specific options, listener options, etc.

Reply:
@mattklein123 Thanks for reviewing.
  Looking at the original code(master), the two constructors differ in - 
- the constructor that takes `fd` does not have an option to `bind`
- only the constructor that takes `address` sets up `SO_REUSEADDR`. 

So wanted to keep the logic same here.

Reply:
Hmm, I suspect that's a bug and not a feature. Can you spend a little bit of time trying to figure out where the constructors are used? I would prefer we take this opportunity to unify the logic if possible?

Reply:
I think the idea is that when an `fd` is present, then we don't create the listening socket and assume that the socket is already bound (to that listen `fd`). So in that case, we should not be applying the option `SO_REUSEADDR`. All other options could still be applied . Although its confusing to have flag `PRE_BIND` for all the options.

So maybe we should rename `setProtocolSpecificSocketOptions` to `setPrebindSocketOptions` ?

Reply:
That makes sense, +1 please rename.